### DISCUSSION TITLE
Grouping Logs in Screen Flows Across Multiple Screens

### DISCUSSION THREAD

**Discussion by KarciSF** [jongpie/NebulaLogger#625] [Q&A]
Hello guys,

I'm currently exploring the functionality of Nebula Logger and have encountered a specific challenge related to screen flows. My objective is to group logs from a screen flow that is segmented by different screens (and thus breaking transactions) into a single consolidated log.

Here's what I've tried so far:

1. **Delaying Log Entries:** Initially, I attempted to prevent log entries from being saved at each step. Instead, I opted to save the log only at the end of the entire flow. However, this approach resulted in only the log from the last screen (up to the 'Save Log' element) being saved in the database, meaning not all logs were captured.
2. **Saving Logs for Each Entry:** Alternatively, when I tried saving a log for each entry, it led to multiple logs being created for each part of the flow that was divided by any element breaking the transaction. This isn't ideal as it results in numerous logs for a single execution of the flow, therefor it is missing the unified picture of flow execution.

During my research, I came across a tutorial by Matt Gerry where he discussed linking different Aura components together under one session using the platform cache. Could this approach be adapted for use with flows as well, possibly out of the box, or would it require code adjustments to support this functionality?

As someone new to Nebula Logger, I believe I might not be the first to face this challenge. I'm hopeful that there might be a way to achieve this with the current version of Nebula Logger, although I haven't found any specific information or guidelines on this yet.

Any guidance, tips, or shared experiences would be greatly appreciated!

Thank you in advance for your support and assistance.
Karol

**Comment by jongpie** [comment]
Hi @KarciSF - this is great feedback, I don't typically work with Flows myself (most of my work is done in Apex & LWC), so I appreciate the insight on this. And based on the 2 ways you've tried so far, I think the "right" way to handle logging in a Screen Flow is to use your option "Saving Logs for Each Entry" - you'll want to save logs before switching any element that breaks the transaction/starts a new transaction.

However, to your point - this approach is missing a unified picture of the Screen Flow's whole execution. In the current version of Nebula Logger, I don't think there's a way to properly handle this (and I don't think using Platform Cache would scale well), but I think there are a few enhancements that can be made to improve logging in Screen Flows:

- **Linking related `Log__c` records can't be done in Flow (yet)**. The goal of the `Log__c` object is to represent a single transaction in Apex/Flow. But in Salesforce, there are a lot of situations where you want to have context of logging data that spans multiple transactions (Screen Flows, Scheduled Flows, Apex Batchable, Apex Queueable, etc).
  - In Apex, Nebula Logger already provides the ability to link multiple `Log__c` records (transactions) together, using the methods `Logger.setParentTransactionId(String)`, `Logger.getParentTransaction()`, and `Logger.getTransactionId()`. The [README has an example](https://github.com/jongpie/NebulaLogger/tree/v4.12.9?tab=readme-ov-file#track-related-logs-in-batchable-and-queuable-jobs) of using these methods in batch Apex. 
  - In Flow, the logging actions don't have any equivalent functionality. I think this should be a fairly easy enhancement. This will then provide the necessary data points to link the `Log__c` and `LogEntry__c` together from the Screen Flow's multiple transactions. I've created issue #626 to track this enhancement.
- **Viewing all logging data for related `Log__c` records can't be (easily) done (yet)**. There have been a few discussions, including [this comment I wrote last June](https://github.com/jongpie/NebulaLogger/issues/499), about improving the reportability of related `Log__c` records. 
  - I think the existing data model already has the necessary relationships to store the data, but Salesforce's built-in tools (related lists, list views, reports, etc) don't make it easy to visualize this data. 
  - I'm considering creating an LWC to display on the `Log__c` page to provide a cleaner way to see all related `Log__c` and `LogEntry__c` records. I'm still exploring this option, but I think building a custom solution (LWC) will provide the most flexibility.

Let me know what you think. My hope is to make progress on some of these enhancements over the next month (I have a few other enhancements & bugfixes that I need to finish first).

**Reply by KarciSF** [reply]
Hey @jongpie ,
thanks a lot for getting back to me with such a detailed reply! It's super cool to see my challenges with Screen Flows on your radar, and I'm honestly excited about the enhancements you're considering. It sounds like you've got a solid grip on the whole logging puzzle, and extending that Apex magic to Flows should help to tackle this challenge.

I read the other comment (with Stephen) and understand your perspective on `Log__c` usage for batch/queueable jobs. Do you envision the same behavior for screen flows? Even with using `setParentTransactionId`, logs would at least be grouped, but I agree, readability might not be ideal. What about giving option to consolidate logs with the same TransactionId into one? Or is that not a direction you're keen onfollowing? I believe Nebula Logger will be an excellent tool for our organization, and it's exciting to see its continuous evolution.

Thanks again for taking my feedback to heart.

Cheers,
Karol

**Reply by jongpie** [reply]
Hi @KarciSF at the moment, I envision that screen flows will follow the same/current data model used for Apex with `setParentTransactionId`, as combining `Log__c` records would cause some headaches with reporting & other functionality. But as Nebula Logger continues to grow & evolve, I'm frequently reevaluating the data model to see if there are opportunities to improve it, so I'll definitely continue considering how to best handle data that spans multiple transactions.

And I'm really glad to hear you think it'll be an excellent tool for your organization, and thanks so much for all of the feedback you've already provided!

**Reply by KarciSF** [reply]
Introducing the option to use `TransactionId` will indeed solve the issue. The next step for me would be to visualize the entries from all child logs at the top-level log. I'm considering creating a quick Screen Flow using datatable for this purpose, or possibly an LWC.

I also have one more question, which I'll post in a separate post. 😄 I hope I'm not spamming you too much!
